packages/core/contracts/financial-templates/common/FundingRateApplier.sol

```diff
@@ -0,0 +1,213 @@
+
+    modifier updateFundingRate {
+        _applyEffectiveFundingRatePerToken();
+        _;
+    }
+
+    modifier notEmergencyShutdown() {
+        _notEmergencyShutdown();
+        _;
+    }
+
+    modifier isEmergencyShutdown() {
+        _isEmergencyShutdown();
+        _;
+    }
+
+    /**
+     * @notice Constructs the FundingRateApplier contract. Called by child contracts.
+     * @param _initialFundingRate Starting funding rate multiplier.
+     * @param _fpFinderAddress Finder used to discover financial-product-related contracts.
+     * @param _priceIdentifier Unique identifier for DVM price feed ticker for child financial contract.
+     * @param _timerAddress Contract that stores the current time in a testing environment.
+     * Must be set to 0x0 for production environments that use live time.
+     */
+    constructor(
+        FixedPoint.Unsigned memory _initialFundingRate,
+        address _fpFinderAddress,
+        bytes32 _priceIdentifier,
+        address _timerAddress
+    ) public nonReentrant() {
+        cumulativeFundingRateMultiplier = _initialFundingRate;
+        fpFinder = FinderInterface(_fpFinderAddress);
+        priceIdentifier = _priceIdentifier;
+
+        timer = Timer(_timerAddress);
+        lastUpdateTime = timer.getCurrentTime();
+    }
+
+    /****************************************
+     *         PUBLIC FUNCTIONS           *
+     ****************************************/
+
+    /**
+     * @notice Fetches the latest funding rate from the Store, scales it over the time period since the last update,
+     * and uses this effective rate to set a new funding rate multiplier.
+     * @dev A funding rate of 1.0 reported by the Store implies a neutral funding rate, meaning that the current multiplier
+     * should not change.
+     */
+    function applyFundingRate() public {
+        _applyEffectiveFundingRatePerToken();
+    }
+
+    // Returns a token amount scaled by the current funding rate multiplier.
+    // Note: if the contract has paid fees since it was deployed, the raw value should be larger than the returned value.
+    function _getFundingRateAppliedTokenDebt(FixedPoint.Unsigned memory rawTokenDebt)
+        public
+        view
+        returns (FixedPoint.Unsigned memory tokenDebt)
+    {
+        return rawTokenDebt.mul(cumulativeFundingRateMultiplier);
+    }
+
+    /****************************************
+     *         INTERNAL FUNCTIONS           *
+     ****************************************/
+
+    function _getFundingRateStore() internal view returns (FundingRateStoreInterface) {
+        return FundingRateStoreInterface(fpFinder.getImplementationAddress("FundingRateStore"));
+    }
+
+    function _getLatestFundingRate() internal view returns (FixedPoint.Unsigned memory) {
+        FundingRateStoreInterface fundingRateStore = _getFundingRateStore();
+        return fundingRateStore.getLatestFundingRateForIdentifier(priceIdentifier);
```

Comment:
Will the funding rate have a different identifier than the underlying price?

For instance, there may be many tokens that track USDETH price. Each would likely have their own parameters, etc, and each would probably have its own funding rate identifier to determine each token's individual funding rate, right?

Reply:
Yes I think the funding rate should have a diff identifier for the reason you identify. Chris and i discussed this and this [issue](https://github.com/UMAprotocol/protocol/issues/2086) came out of it

Reply:
Yup. this is a good point and I agree that the identifier should be registered with the DVM as a separate identifier to the contracts `priceIdentifier` as UMA token holders will be required to vote on this. I have updated the PR accordingly.